brokers/artifacts/extensions.go

```diff
@@ -0,0 +1,95 @@
+//
+// Copyright (c) 2019 Red Hat, Inc.
+// This program and the accompanying materials are made
+// available under the terms of the Eclipse Public License 2.0
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+//
+
+package artifacts
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"github.com/eclipse/che-plugin-broker/model"
+	"github.com/eclipse/che-plugin-broker/utils"
+)
+
+// ProcessPlugin downloads all undownloaded plugin extensions and places the
+// relevant artifacts in their appropriate location in the /plugins directory.
+// If a plugin already has artifacts downloaded for a given extension, that
+// extension is skipped.
+func (b *Broker) ProcessPlugin(plugin *model.CachedPlugin) error {
+	workDir, err := b.ioUtils.TempDir("", "artifacts-broker")
+	if err != nil {
+		return err
+	}
+
+	var infoLine strings.Builder
+	defer b.PrintInfoBuffer(&infoLine)
+
+	infoLine.WriteString(fmt.Sprintf("Processing plugin %s", plugin.ID))
+	numExtensions := len(plugin.CachedExtensions)
+	extensionIdx := 0
+	for URL, path := range plugin.CachedExtensions {
+		extensionIdx = extensionIdx + 1
+		infoLine.WriteString(fmt.Sprintf("\n  Installing plugin extension %d/%d", extensionIdx, numExtensions))
+		if path != "" {
+			infoLine.WriteString(fmt.Sprintf("\n    Plugin already downloaded"))
+			continue
+		}
+		infoLine.WriteString(fmt.Sprintf("\n    Downloading plugin from %s", URL))
```

Comment:
With it, we get ordered output but we see nothing when something is happening, like downloading...

Maybe it would be better if we flush buffer each time before the time-consuming task is started?

like:
```go
        var infoLine strings.Builder
	defer b.PrintInfoBuffer(&infoLine)

	infoLine.WriteString(fmt.Sprintf("Processing plugin %s", plugin.ID))
	numExtensions := len(plugin.CachedExtensions)
	extensionIdx := 0
	for URL, path := range plugin.CachedExtensions {
		extensionIdx = extensionIdx + 1
		infoLine.WriteString(fmt.Sprintf("\n  Installing plugin extension %d/%d", extensionIdx, numExtensions))
		if path != "" {
			infoLine.WriteString(fmt.Sprintf("\n    Plugin already downloaded"))
			continue
		}
		infoLine.WriteString(fmt.Sprintf("\n    Downloading plugin from %s", URL))

                b.PrintInfoBuffer(&infoLine)
                infoLine.reset()
                ...
```
WDYT?

Reply:
Another implementation I can imagine with channels, where one thread receives messages and collects in into buffer which is flushed like after some period, like 1-2 seconds.
But it would require even more changes in this PR, I'm OK to go with the current implementation.

Reply:
I agree that not showing anything while downloading is a problem, this is mostly a workaround to preserve logs in case of failure. With a slower internet connection, it does look strange, though. 

I like your suggestion of attempting to dump info more often, since as it's currently implemented the java plugin logs *both* extensions at once, which isn't ideal. The main goal is to keep the `downloading plugin x/#` logs for each extension together to avoid interleaving.

I've pushed a commit to rework this to log on every loop iteration and also move to using `[]string` in order to avoid worrying about newlines.